janus_server/src/hpke.rs

```diff
@@ -0,0 +1,379 @@
+    Role, TaskId,
+};
+use hpke::{
+    aead::{Aead, AesGcm256, ChaCha20Poly1305},
+    kdf::{HkdfSha256, HkdfSha512, Kdf},
+    kem::{DhP256HkdfSha256, X25519HkdfSha256},
+    setup_receiver, setup_sender, Deserializable, HpkeError, Kem, OpModeR, OpModeS, Serializable,
+};
+use rand::thread_rng;
+
+#[derive(Debug, thiserror::Error)]
+enum Error {
+    /// Wrapper around errors from crate hpke. See [`hpke::HpkeError`] for more
+    /// details on possible variants.
+    #[error("HPKE error")]
+    Hpke(#[from] HpkeError),
+    #[error("invalid HPKE configuration: {0}")]
+    InvalidConfiguration(&'static str),
+}
+
+/// Labels incorporated into HPKE application info string
+#[derive(Clone, Copy, Debug)]
+enum Label {
+    InputShare,
+    AggregateShare,
+}
+
+impl Label {
+    fn as_bytes(&self) -> &'static [u8] {
+        match self {
+            Self::InputShare => "ppm input share",
+            Self::AggregateShare => "ppm aggregate share",
+        }
+        .as_bytes()
+    }
+}
+
+/// PPM protocol message representing an HPKE private key, serialized using the
+/// `SerializePrivateKey` function as described in RFC 9180, §4 and §7.1.2.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct HpkePrivateKey(pub(crate) Vec<u8>);
+
+impl HpkePrivateKey {
+    pub(crate) fn new(bytes: Vec<u8>) -> Self {
+        Self(bytes)
+    }
+}
+
+impl AsRef<[u8]> for HpkePrivateKey {
+    fn as_ref(&self) -> &[u8] {
+        self.0.as_ref()
+    }
+}
+
+impl From<Vec<u8>> for HpkePrivateKey {
+    fn from(v: Vec<u8>) -> Self {
+        Self::new(v)
+    }
+}
+
+/// Application info used in HPKE context construction
+#[derive(Clone, Debug)]
+struct HpkeApplicationInfo(Vec<u8>);
+
+impl HpkeApplicationInfo {
+    /// Construct HPKE application info from the provided PPM task ID, label and
+    /// participant roles.
+    fn new(task_id: TaskId, label: Label, sender_role: Role, recipient_role: Role) -> Self {
+        Self(
+            [
+                task_id.as_bytes(),
+                label.as_bytes(),
+                &[sender_role as u8],
+                &[recipient_role as u8],
+            ]
+            .concat(),
+        )
+    }
+}
+
+/// A one-shot HPKE sender that encrypts messages to the public key in
+/// `recipient_config` using the AEAD, key derivation and key encapsulation
+/// mechanisms specified in `recipient_config`, and using `label`, `sender_role`
+/// and `recipient_role` to derive application info.
+//
+// This type only exists separately from Sender so that we can have a type that
+// doesn't "leak" the generic type parameters into the caller.
+struct HpkeSender {
+    task_id: TaskId,
+    recipient_config: HpkeConfig,
+    label: Label,
+    sender_role: Role,
+    recipient_role: Role,
+}
+
+impl HpkeSender {
+    /// Encrypt `plaintext` and return the HPKE ciphertext.
+    fn seal(self, plaintext: &[u8], associated_data: &[u8]) -> Result<HpkeCiphertext, Error> {
```

Comment:
As with `HpkeReceiver::open()`, this could take self by reference, as the bare function `seal()` ensures we don't use a context more than once.

Reply:
Done!